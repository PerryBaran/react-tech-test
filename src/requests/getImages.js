import axios from "axios";

const getImages = async (query) => {
  if (!query) return Promise.resolve([]);
  try {
    const response = await axios.get(`https://images-api.nasa.gov/search?q=${query}`);
    const imageResults = response.data.collection.items;
    const parsedImages = imageResults.filter((image) => {
      return image.data[0].media_type === "image";
    });
    const images = parsedImages.map((image) => {
      return image.links[0].href;
    });
    return images;
  } catch (err) {
    console.error(err);
  }
};

export default getImages;