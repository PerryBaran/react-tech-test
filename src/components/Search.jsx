import React, { useRef } from "react";
import PropTypes from "prop-types";
import "../styles/search.css";
import getImages from "../requests/getImages";

function Search({setSearchResults}) {
  const searchRef = useRef();

  const handleSubmit = async (e) => {
    e.preventDefault();
    const images = await getImages(searchRef.current.value);
    setSearchResults(images);
  };

  return (
    <form className="search-form" onSubmit={handleSubmit}>
      <input type="text" className="search-input" ref={searchRef} />
      <button type="submit" className="search-submit">search</button>
    </form>
  );
}

Search.propTypes = {
  setSearchResults: PropTypes.func.isRequired,
};

export default Search;
