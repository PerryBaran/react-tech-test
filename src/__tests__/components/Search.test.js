import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import Search from "../../components/Search";
import * as getImages from "../../requests/getImages";

describe("Search", () => {
  let setSearchResults;

  beforeEach(() => {
    setSearchResults = jest.fn();
  });

  test("snapshot", () => {
    const { asFragment } = render(
      <Search setSearchResults={setSearchResults} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders correctly", () => {
    render(<Search setSearchResults={setSearchResults} />);
    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    expect(input).toBeInTheDocument();
    expect(button).toHaveTextContent("search");
    expect(button).toHaveAttribute("type", "submit");
  });

  test("search functionality", async () => {
    const mockReturn = "return";
    const string = "search";
    jest.spyOn(getImages, "default").mockResolvedValue(mockReturn);
    render(<Search setSearchResults={setSearchResults} />);
    const input = screen.getByRole("textbox");
    const button = screen.getByRole("button");

    expect(getImages.default).toBeCalledTimes(0);
    expect(setSearchResults).toBeCalledTimes(0);
    fireEvent.change(input, { target: { value: string } });
    fireEvent.click(button);
    await waitFor(() => expect(getImages.default).toBeCalledWith(string));
    await waitFor(() => expect(setSearchResults).toBeCalledWith(mockReturn));
  });
});
