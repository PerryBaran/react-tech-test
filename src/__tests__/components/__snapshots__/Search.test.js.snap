// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search snapshot 1`] = `
<DocumentFragment>
  <form
    class="search-form"
  >
    <input
      class="search-input"
      type="text"
    />
    <button
      class="search-submit"
      type="submit"
    >
      search
    </button>
  </form>
</DocumentFragment>
`;
